File: src/dere_discord/persona.py
"""Persona loading utilities for dere-discord."""

from __future__ import annotations

from collections.abc import Iterable
from dataclasses import dataclass

from dere_shared.personalities import PersonalityLoader

from .paths import get_config_dir


@dataclass(slots=True)
class PersonaProfile:
    """Persona metadata for downstream consumers."""

    names: tuple[str, ...]
    prompt: str
    color: str | None
    icon: str | None


class PersonaService:
    """Resolve personality combinations into prompts and metadata."""

    def __init__(self, default_personas: tuple[str, ...]):
        self._default_personas = default_personas
        self._loader = PersonalityLoader(get_config_dir())
        self._identity: str | None = None

    @property
    def default_personas(self) -> tuple[str, ...]:
        return self._default_personas

    def set_identity(self, identity: str | None) -> None:
        self._identity = identity

    def resolve(self, personas: Iterable[str] | None = None) -> PersonaProfile:
        """Resolve persona tuple into prompt + display metadata."""

        names = tuple(personas) if personas else self._default_personas
        prompts: list[str] = []
        color: str | None = None
        icon: str | None = None

        for idx, name in enumerate(names):
            personality = self._loader.load(name)
            if idx == 0:
                color = personality.color
                icon = personality.icon
            prompts.append(personality.prompt_content)

        prompt_text = "\n\n".join(prompts)
        if self._identity:
            prompt_text = (
                f"You are {self._identity}. Unless asked otherwise, introduce yourself "
                "using this name.\n\n"
            ) + prompt_text
        return PersonaProfile(names=names, prompt=prompt_text, color=color, icon=icon)

File: packages/discord/src/persona.ts
import { parse } from "@iarna/toml";
import { readFile } from "node:fs/promises";
import { basename, dirname, join } from "node:path";
import { getConfigPath } from "@dere/shared-config";

export type PersonaProfile = {
  names: string[];
  prompt: string;
  color: string | null;
  icon: string | null;
};

type PersonalityDoc = {
  metadata?: {
    name?: string;
    short_name?: string;
    aliases?: string[];
  };
  display?: {
    color?: string;
    icon?: string;
    avatar?: string;
    announcement?: string;
  };
  prompt?: {
    content?: string;
  };
};

type Personality = {
  name: string;
  short_name: string;
  aliases: string[];
  color: string;
  icon: string;
  prompt_content: string;
  announcement?: string;
};

function defaultEmbeddedDir(): string {
  const envDir = process.env.DERE_EMBEDDED_PERSONALITIES_DIR;
  if (envDir) {
    return envDir;
  }
  return join(process.cwd(), "packages", "shared-assets", "personalities");
}

function defaultUserDir(): string {
  return join(dirname(getConfigPath()), "personalities");
}

function parsePersonality(data: string): Personality {
  const parsed = parse(data) as PersonalityDoc;
  const metadata = parsed.metadata ?? {};
  const display = parsed.display ?? {};
  const prompt = parsed.prompt ?? {};

  return {
    name: metadata.name ?? "",
    short_name: metadata.short_name ?? "",
    aliases: metadata.aliases ?? [],
    color: display.color ?? "white",
    icon: display.icon ?? "*",
    prompt_content: prompt.content ?? "",
    announcement: display.announcement,
  };
}

async function loadPersonalityFromFile(path: string): Promise<Personality> {
  const text = await readFile(path, "utf-8");
  return parsePersonality(text);
}

export class PersonalityLoader {
  private readonly userDir: string;
  private readonly embeddedDir: string;

  constructor(configDir: string) {
    this.userDir = join(configDir, "personalities");
    this.embeddedDir = defaultEmbeddedDir();
  }

  async load(name: string): Promise<Personality> {
    const userPath = join(this.userDir, `${name}.toml`);
    try {
      return await loadPersonalityFromFile(userPath);
    } catch {
      const embeddedPath = join(this.embeddedDir, `${name}.toml`);
      return await loadPersonalityFromFile(embeddedPath);
    }
  }
}

export class PersonaService {
  private readonly defaultPersonas: string[];
  private readonly loader: PersonalityLoader;
  private identity: string | null = null;

  constructor(defaultPersonas: string[]) {
    this.defaultPersonas = defaultPersonas;
    this.loader = new PersonalityLoader(dirname(getConfigPath()));
  }

  get defaults(): string[] {
    return this.defaultPersonas;
  }

  setIdentity(identity: string | null): void {
    this.identity = identity;
  }

  async resolve(personas?: Iterable<string> | null): Promise<PersonaProfile> {
    const names = personas ? Array.from(personas) : this.defaultPersonas;
    const prompts: string[] = [];
    let color: string | null = null;
    let icon: string | null = null;

    for (let idx = 0; idx < names.length; idx += 1) {
      const name = names[idx] ?? "";
      if (!name) {
        continue;
      }
      const personality = await this.loader.load(name);
      if (idx === 0) {
        color = personality.color;
        icon = personality.icon;
      }
      prompts.push(personality.prompt_content);
    }

    let promptText = prompts.join("\n\n");
    if (this.identity) {
      promptText =
        `You are ${this.identity}. Unless asked otherwise, introduce yourself using this name.\n\n` +
        promptText;
    }

    return {
      names: names.filter(Boolean),
      prompt: promptText,
      color,
      icon,
    };
  }
}

export function normalizePersonaName(name: string): string {
  return basename(name, ".toml");
}
